### PR TITLE
*: support adding user in v3 auth

### PR DESCRIPTION
**Pull request by mitake**
This commit adds a new subcommand "user add" to etcdctlv3. With the
command users can create a user for the authentication.

Example of usage:
$ etcdctlv3 user add user1
Password of user1:
Type password of user1 again for confirmation:

### CODE REVIEW

File: etcdserver/etcdserverpb/raft_internal.proto [hunk 1]
```diff
@@ -26,6 +26,7 @@ message InternalRaftRequest {
   LeaseRevokeRequest lease_revoke = 9;
 
   AuthEnableRequest auth_enable = 10;
+  UserAddRequest user_add = 11;
```
Comment (xiang90) [inline etcdserver/etcdserverpb/raft_internal.proto]: AuthAddUserRequest auth_user_add = 11;

File: etcdserver/etcdserverpb/rpc.proto [hunk 2]
```diff
@@ -494,6 +496,13 @@ message AuthenticateResponse {
 
 message UserAddResponse {
   ResponseHeader header = 1;
+  enum Result {
```
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.proto]: we do not need this actually. we can return error on error case. We can return duplicate error or exhausted error.

File: etcdserver/api/v3rpc/auth.go [hunk 3]
```diff
@@ -68,8 +68,7 @@ func (as *AuthServer) RoleGrant(ctx context.Context, r *pb.RoleGrantRequest) (*p
 }
 
 func (as *AuthServer) UserAdd(ctx context.Context, r *pb.UserAddRequest) (*pb.UserAddResponse, error) {
-	plog.Info("not implemented yet")
-	return nil, nil
+	return as.authenticator.UserAdd(ctx, r)
```
Comment (xiang90) [inline etcdserver/api/v3rpc/auth.go]: add the duplicate error and exhausted error into v3rpc/errors.go. then we can use gRPC to return the error.

File: auth/store.go [hunk 4]
```diff
@@ -56,10 +78,105 @@ func (as *authStore) Recover(be backend.Backend) {
...
+			}
+
+			return i*8 + int(j)
+		}
+	}
+
+	return -1
+}
+
+func (as *authStore) UserAdd(r *pb.UserAddRequest) (*pb.UserAddResponse, error) {
```
Comment (xiang90) [inline auth/store.go]: I feel the ID generation is overly complicated. Is there a reason that we cannot just use user name as its ID? So the key in the authStore will be the username and value is the proto struct.

File: etcdserver/etcdserverpb/raft_internal.proto (hunk 1 above)
Comment (mitake) [inline etcdserver/etcdserverpb/raft_internal.proto]: OK. Should I change the type name, too (including etcdserver/etcdserverpb/rpc.proto)?

File: etcdserver/api/v3rpc/auth.go (hunk 3 above)
Comment (mitake) [inline etcdserver/api/v3rpc/auth.go]: I see.

File: auth/store.go (hunk 4 above)
Comment (mitake) [inline auth/store.go]: At first I thought using a user name as a key would make a problem in this kind of situation: create user A -> delete A -> create A (same name). However, the user name isn't refereed by other data structure so using it directly as a key would be fine. I'll fix it.

(I think role should use numerical ID. It is refereed from the user struct so recycling name introduce the problem).

**Comment by mitake** [comment]
@xiang90 updated the parts of errocode and user ID. How should I change the line `UserAddRequest user_add = 11;`? Should I change its type name, too?

File: auth/authpb/auth.proto [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_getters_all) = false;
+option (gogoproto.goproto_enum_prefix_all) = false;
+
+// User is a single entry in the bucket authUsers
+message User {
+  bytes password = 1;
```
Comment (xiang90) [inline auth/authpb/auth.proto]: i feel duplicating userName in this struct is fine. it makes the struct more clear.

File: auth/store.go [hunk 6]
```diff
@@ -15,23 +15,40 @@
...
-	authBucketName = []byte("auth")
+	enableFlagKey       = []byte("authEnabled")
+	authBucketName      = []byte("auth")
+	authUsersBucketName = []byte("authUsers")
 
 	plog = capnslog.NewPackageLogger("github.com/coreos/etcd", "auth")
 )
 
+// FIXME: copied from lease, need to unify?
+func int64ToBytes(n int64) []byte {
```
Comment (xiang90) [inline auth/store.go]: remove this? no one uses this func now.

File: auth/store.go [hunk 7]
```diff
@@ -15,23 +15,40 @@
...
+}
+
 type AuthStore interface {
 	// AuthEnable() turns on the authentication feature
 	AuthEnable()
 
 	// Recover recovers the state of auth store from the given backend
 	Recover(b backend.Backend)
+
+	// UserAdd() adds a new user
```
Comment (xiang90) [inline auth/store.go]: UserAdd adds a new user. remove "()"

File: auth/store.go [hunk 8]
```diff
@@ -56,10 +73,53 @@ func (as *authStore) Recover(be backend.Backend) {
 	// TODO(mitake): recovery process
 }
 
+func (as *authStore) UserAdd(r *pb.UserAddRequest) (*pb.UserAddResponse, error) {
+	plog.Infof("adding a new user: %s", r.Name)
```
Comment (xiang90) [inline auth/store.go]: noticef?

File: auth/store.go [hunk 9]
```diff
@@ -56,10 +73,53 @@ func (as *authStore) Recover(be backend.Backend) {
...
+		plog.Errorf("failed to hash password: %s", err)
+		return nil, err
+	}
+
+	tx := as.be.BatchTx()
+	tx.Lock()
+
+	_, vs := tx.UnsafeRange(authUsersBucketName, []byte(r.Name), nil, 0)
+	if len(vs) != 0 {
+		plog.Errorf("User name %s is already used", r.Name)
```
Comment (xiang90) [inline auth/store.go]: release tx.lock?

Comment (xiang90) [inline auth/store.go]: i do not think we should log user error at server side.

File: auth/store.go [hunk 10]
```diff
@@ -56,10 +73,53 @@ func (as *authStore) Recover(be backend.Backend) {
...
+		plog.Errorf("failed to marshal a new user data: %s", merr)
+		tx.Unlock()
+
+		return nil, merr
+	}
+
+	tx.UnsafePut(authUsersBucketName, []byte(r.Name), marshaledUser)
+
+	tx.Unlock()
+	as.be.ForceCommit()
```
Comment (xiang90) [inline auth/store.go]: we do not need to do forceCommit here i believe.

Comment (xiang90) [inline auth/store.go]: we only do force commit before transferring a snapshot.

File: auth/store.go [hunk 11]
```diff
@@ -56,10 +73,53 @@ func (as *authStore) Recover(be backend.Backend) {
...
+
+	plog.Infof("added a new user: %s", r.Name)
```
Comment (xiang90) [inline auth/store.go]: noticef?

File: etcdctlv3/command/user_command.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+	}
+}
+
+// userAddCommandFunc executes the "user add" command.
+func userAddCommandFunc(cmd *cobra.Command, args []string) {
+	if len(args) != 1 {
+		ExitWithError(ExitBadArgs, fmt.Errorf("user add command requires user name as its argument."))
+	}
+
+	prompt1 := fmt.Sprintf("Password of %s: ", args[0])
```
Comment (xiang90) [inline etcdctlv3/command/user_command.go]: i think @heyitsanthony wants this to accept password from stdin.

File: etcdctlv3/command/user_command.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	password1, err1 := speakeasy.Ask(prompt1)
+	if err1 != nil {
+		ExitWithError(ExitBadArgs, fmt.Errorf("failed to ask password: %s.", err1))
+	}
+
+	if len(password1) == 0 {
+		ExitWithError(ExitBadArgs, fmt.Errorf("Empty password."))
```
Comment (xiang90) [inline etcdctlv3/command/user_command.go]: empty password

File: etcdserver/api/v3rpc/rpctypes/error.go [hunk 14]
```diff
@@ -35,4 +35,6 @@ var (
 	ErrMemberNotFound = grpc.Errorf(codes.NotFound, "etcdserver: member not found")
 
 	ErrRequestTooLarge = grpc.Errorf(codes.InvalidArgument, "etcdserver: request is too large")
+
+	ErrUserNameAlreadyUsed = grpc.Errorf(codes.FailedPrecondition, "etcdserver: user name is already used")
```
Comment (xiang90) [inline etcdserver/api/v3rpc/rpctypes/error.go]: ErrorUserAlreadyExist

**Comment by xiang90** [comment]
>  How should I change the line UserAddRequest user_add = 11;? Should I change its type name, too?

I hope it can have the auth prefix. So AuthUserAddRequest auth_user_add instead of just user_add.

File: etcdctlv3/command/user_command.go (hunk 12 above)
Comment (heyitsanthony) [inline etcdctlv3/command/user_command.go]: The problem with speakeasy is it expects an interactive terminal, which makes it difficult to write e2e tests. (cf. 02d9aa481bc5e689399d0ae841cfe7ace0d252c1)

File: auth/authpb/auth.proto (hunk 5 above)
Comment (mitake) [inline auth/authpb/auth.proto]: OK

File: auth/store.go (hunk 6 above)
Comment (mitake) [inline auth/store.go]: Oops, thanks for pointing. I'll delete it in the next update.

File: auth/store.go (hunk 7 above)
Comment (mitake) [inline auth/store.go]: OK, thanks for pointing

File: auth/store.go (hunk 8 above)
Comment (mitake) [inline auth/store.go]: OK. BTW, is there a guideline of log level?

File: auth/store.go (hunk 9 above)
Comment (mitake) [inline auth/store.go]: OK

File: auth/store.go (hunk 10 above)
Comment (mitake) [inline auth/store.go]: I misunderstood the API, will fix in the next update. Thanks.

File: auth/store.go (hunk 11 above)
Comment (mitake) [inline auth/store.go]: OK

File: etcdctlv3/command/user_command.go (hunk 12 above)
Comment (mitake) [inline etcdctlv3/command/user_command.go]: OK, I'll prepare the feature for reading password from stdin.

File: etcdctlv3/command/user_command.go (hunk 13 above)
Comment (mitake) [inline etcdctlv3/command/user_command.go]: OK

File: etcdserver/api/v3rpc/rpctypes/error.go (hunk 14 above)
Comment (mitake) [inline etcdserver/api/v3rpc/rpctypes/error.go]: OK

**Comment by mitake** [comment]
@xiang90 I see. Then I'll add the Auth prefix to the type name.

File: auth/store.go (hunk 8 above)
Comment (xiang90) [inline auth/store.go]: no. probably we should add.

**Comment by mitake** [comment]
@xiang90 updated this PR except renaming `UserAddRequest`. Because there are other messages that don't follow the auth prefix, I'd like to update them in a single separated PR for more readable git log. How do you think? Or should I change the name in this PR?

@heyitsanthony password can be read from stdin with the option `--password-stdin`.

File: auth/store.go [hunk 15]
```diff
@@ -56,10 +64,53 @@ func (as *authStore) Recover(be backend.Backend) {
...
+	plog.Noticef("adding a new user: %s", r.Name)
+
+	hashed, err := bcrypt.GenerateFromPassword([]byte(r.Password), bcrypt.DefaultCost)
+	if err != nil {
+		plog.Errorf("failed to hash password: %s", err)
+		return nil, err
+	}
+
+	tx := as.be.BatchTx()
+	tx.Lock()
```
Comment (xiang90) [inline auth/store.go]: how about just defer tx.Unlock()?

File: etcdserver/api/v3rpc/rpctypes/error.go [hunk 16]
```diff
@@ -35,4 +35,6 @@ var (
 	ErrMemberNotFound = grpc.Errorf(codes.NotFound, "etcdserver: member not found")
 
 	ErrRequestTooLarge = grpc.Errorf(codes.InvalidArgument, "etcdserver: request is too large")
+
+	ErrUserAlreadyExist = grpc.Errorf(codes.FailedPrecondition, "etcdserver: user name is already used")
```
Comment (xiang90) [inline etcdserver/api/v3rpc/rpctypes/error.go]: change the error message too? user name already exists.

File: etcdserver/api/v3rpc/auth.go (hunk 3 above)
Comment (xiang90) [inline etcdserver/api/v3rpc/auth.go]: we need to check the error here and convert it to the `ErrUserAlreadyExist ` error defined in error.go.

**Comment by xiang90** [comment]
Looks good to me after fixing the last few issues.

/cc @heyitsanthony can you also give this a look (especially the etcdctl part)

File: etcdserver/api/v3rpc/auth.go (hunk 3 above)
Comment (mitake) [inline etcdserver/api/v3rpc/auth.go]: Do you mean `auth` package shouldn't use the `rpctypes` directly?

**Comment by mitake** [comment]
@xiang90 updated expect the error handling in `etcdserver/api/v3rpc/auth.go`. Should I prepare an internal error code for the auth errors?

**Comment by mitake** [comment]
@xiang90 Or do you mean we need to convert errors from `bcrypt.GenerateFromPassword()`?

**Comment by xiang90** [comment]
@mitake When will bcrypt generate an error?

File: etcdserver/api/v3rpc/auth.go (hunk 3 above)
Comment (xiang90) [inline etcdserver/api/v3rpc/auth.go]: At rpc level, we have to convert all errors into the error defined by the rpc layer. Or the client side will not be able to parse the error.

**Comment by mitake** [comment]
@xiang90 I'm not sure but the type of the function can return the error.

**Comment by xiang90** [comment]
@mitake Probably just add a unknow type error into errors.go. And we check the error returned by auth, if it is an unknown one, just return as error message (etcdserver: unknow error %v, error) and grpc.Unknown as its type.

**Comment by xiang90** [comment]
@mitake I can clean up the error code thing later. the code lgtm. i just want to wait for @heyitsanthony to review the etcdctl part. thanks!

**Comment by mitake** [comment]
@xiang90 I understand, thanks for your review!

File: etcdctlv3/command/user_command.go [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+)
+
+func NewUserAddCommand() *cobra.Command {
+	cmd := cobra.Command{
+		Use:   "add <user name>",
+		Short: "add a new user",
+		Run:   userAddCommandFunc,
+	}
+
+	cmd.Flags().BoolVar(&readPasswordFromStdin, "password-stdin", false, "read password from stdin instead of interactive terminal")
```
Comment (heyitsanthony) [inline etcdctlv3/command/user_command.go]: I think `interactive` (default true) would be better than `password-stdin`.

Ideally this would use something like `isatty` to autodetect instead of depending on command line flag, but that can be a separate PR since it means pulling in a new godep. A flag is probably good enough anyway. @xiang90, thoughts?

**Comment by heyitsanthony** [comment]
lgtm aside from the flag name. Thanks!

File: etcdctlv3/command/user_command.go (hunk 17 above)
Comment (xiang90) [inline etcdctlv3/command/user_command.go]: i agree. let's change this to interactive for now.

**Comment by mitake** [comment]
@heyitsanthony @xiang90 renamed the flag, PTAL.
